File: 1477-product-of-the-last-k-numbers/product-of-the-last-k-numbers.java
class ProductOfNumbers {

    int lz;
    List<Integer> arr;

    public ProductOfNumbers() 
    {
        arr=new ArrayList<>();
        arr.add(1);
    }
    
    public void add(int num) 
    {
        int index= arr.size()-1;

        if(num==0)
        lz=index+1;

        if(arr.get(index)==0)
        arr.add(num);

        else
        arr.add(arr.get(index)*num);
    }
    
    public int getProduct(int k) 
    {
        int index= arr.size()-1;
        
        if(index-k<lz)
        return 0;

        if(arr.get(index-k)==0)
        return arr.get(index);

        return arr.get(index)/arr.get(index-k);
    }
}

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers obj = new ProductOfNumbers();
 * obj.push(num);
 * int param_2 = obj.getProduct(k);
 */

File: 1477-product-of-the-last-k-numbers/product-of-the-last-k-numbers.cpp
class ProductOfNumbers {
public:
    int lastZeroIndex;
    vector<int> prod;
    ProductOfNumbers() 
    {
        lastZeroIndex = -1;
    }
    
    void add(int num) 
    {
        if(prod.empty() || prod.back() == 0 || num == 0)
        {
            prod.push_back(num);
            // if the number is zero than update the zero index
            if(num == 0)
            lastZeroIndex = prod.size() - 1;

            return;
        }

        // multiply the num with the last entry
        prod.push_back(num * prod.back());
    }
    
    int getProduct(int k) 
    {
        int index = prod.size() - 1;
        // if a zero is present in the k elements
        if(lastZeroIndex > index - k)
        return 0;

        // the zero is present just before last k elements
        if(lastZeroIndex == index - k)
        return prod.back();

        return prod.back() / prod[index - k];
    }
};

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers* obj = new ProductOfNumbers();
 * obj->add(num);
 * int param_2 = obj->getProduct(k);
 */